File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        islands = 0
        def bfs(r, c):
            que = deque()
            que.append((r,c))
            grid[r][c] = '0'
            while que:
                (fr, fc) = que.popleft()
                for dr, dc in [(1,0), (-1,0), (0,1), (0,-1)]:
                    nr = fr + dr
                    nc = fc + dc
                    if nr < 0 or nc < 0 or nr >= ROWS or nc >= COLS or grid[nr][nc] == '0':
                        continue
                    grid[nr][nc] = '0'
                    que.append((nr, nc))

        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == '1':
                    bfs(i, j)
                    islands += 1

        return islands

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int islands = 0;
        Queue<int[]> queue = null;
        int[][] directions = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    islands++;                    
                    queue = new LinkedList<>();
                    grid[i][j] = '0';
                    queue.offer(new int[]{i,j});
                    while (!queue.isEmpty()) {
                        int[] curr = queue.poll();                        
                        for (int[] dir : directions) {
                            int next_i = curr[0] + dir[0];
                            int next_j = curr[1] + dir[1];
                            if (next_i < 0 || next_j < 0 || next_i >= grid.length || next_j >= grid[0].length 
                            || grid[next_i][next_j] == '0') continue;
                            grid[next_i][next_j] = '0';
                            queue.offer(new int[]{next_i, next_j});
                        }
                    }
                }
            }
        }
        return islands;
    }

}